kernel/idle.c

```diff
@@ -164,6 +164,20 @@ void idle(void *unused1, void *unused2, void *unused3)
 	__idle_time_stamp = (u64_t)k_cycle_get_32();
 #endif
 
+#ifdef CONFIG_SMP
+	/* Simplified idle for non-default SMP CPUs pending driver
+	 * support.  The busy waiting is needed to prevent lock
+	 * contention.  Long term we need to wake up idle CPUs with an
+	 * IPI.
+	 */
+	if (_arch_curr_cpu()->id > 0) {
+		while (1) {
+			k_busy_wait(100);
+			k_yield();
+		}
```

Comment:
this means, scheduler code always run on main cpu? also as per the current logic other cpu's don't don't enter sleep states?

Reply:
Timeout code always runs on the main CPU, yes.  Right now it's based on counting ticks instead of finer-grained timeouts, so there's no easy way to translate that to a different CPU's interrupts, which happen at a different phase.   Scheduler code runs on every CPU, it's SMP :)

And yes: right now auxiliary CPUs can't enter a low power state because there is nothing to wake them up appropriately when a new thread enters the scheduler.  We have to choose either bad latency (i.e. have them wake up at their next timer tick interrupt, which can be millions of cycles late) or bad power behavior (spin on the scheduler looking for updates), and I picked the latter.

In the intermediate term we need an architecture where adding a schedulable thread to the run queue where there are idle CPUs will produce an IPI to wake up the others CPUs.

Reply:
got it. Thanks.